PERFORMANCE.md

```diff
@@ -0,0 +1,85 @@
+
+## Performance Sensitive Areas
+
+- Change detection: `observer`, `computed`
+- View creation: `View`, `Renderer`
+- Routing: `Router`
+- Dependency Injection: `Container`, `Registry`
+
+__Note :__ If you are not in a performance sensitive area feel free to use these constructs. That being said, please do your due diligence to ensure your contribution is indeed not in a hot path.
+
+## Array Methods
+
+Most of the time using array methods e.g. `forEach`, `map`, `filter`, is fine, but in hot paths these methods can become very costly. One would think that these methods would almost be like macros that would compile into more efficient code, but that is not the case. In hot paths these methods can be up to 10X slower than just writing a `for`-loop.
+
+One of the reasons behind the slowness is largely do to the fact that these methods take a closure which allocates memory. In turn this creates GC pressure and potentially creates a polymorphic call site. For a more in depth explanation please see the following [explainer](https://gist.github.com/stefanpenner/376efb906388954146a8).
+
+## Closures
+
+Try to keep closures to a minimum in hot code paths.
+
+- They create GC pressure.
+- They may result in polymorphic call site (see below).
+
+The reason behind why closures are slower is that contrary to popular belief, closures actually cause the VM to create a stateful `context` object at the call site. This is very similar to `new`-ing a constructor and creating an instance at the call site. Because of this closures are grouped into the same Polymorphic and Monomorphic call site heuristics.
+
+## Polymorphic vs Monomorphic Call Sites
+
+Calling `fn()` may be fast or slow depending whether the `fn` variable is polymorphic (changes over time) or monomorphic (stable). The difference can be 4x in speed. Monomorphic sites can be inlined, where as polymorphic sites prevent inline and may even cause de-optimization of your code. In general stick to classes with instance method, and avoid polymorphism for code sensitive paths such as change-detection and dependency injection. For more information about how VMs handle optimizations please read [this article](http://mrale.ph/blog/2015/01/11/whats-up-with-monomorphism.html).
+
+## Object Shaping And Type Stability
+
+VMs optimize JS objects by creating hidden classes for each property of that object or context. Because of this you should avoid mutating the shape of an object once it is created.
+
+```
+class Point { // Hidden class 0
+  constructor(x) {
+    this.x = x; // Hidden class 1
+  }
+}
+
+let p1 = new P(10);
+let p2 = new P(11); // Shares hidden class 1
+let p3 = new P(12);
+p3.y = 20; // Creates hidden class 2 :(
+```
+
+The same rules apply for types.
+
+```
+class Point { // Hidden class 0
+  constructor(x) {
+    this.x = x; // Hidden class 1
+  }
+}
+
+let p1 = new P(10);
+let p2 = new P(11); // Shares hidden class 1
+let p3 = new P('12'); // Creates hidden class 2 :(
```

Comment:
I don't believe this creates a new hidden class, although code that expects x to be a number, will now discover it not to be.